[ENT-676] Degreed Integration

UmanShahzad opened the pull request:
# Degreed Integration

This PR adds a new integration channel, namely Degreed.

It necessarily refactors a lot of the existing code infrastructure available for integrated channels. It makes it easier to extend or override base integration channel classes and methods to add new, specific integration channels, like for Degreed.

The structure that this PR aims to give integration channels overall, to be followed by any new integration channel, is the following:

```
Integration Channel
    - Exporters
        - Learner Data
        - Course Metadata
    - Transmitters
        - Learner Data
        - Course Metadata
    - API Client
    - Models
        - Global Integration Channel Configuration
        - Integration Channel Enterprise Customer Configuration
        - Catalog Data Transmission Audit
        - Learner Data Transmission Audit
```

## Exporters

An **exporter** is essentially a **serializer** (I would have gone with this name but we were already using 'Exporter' so I decided to go with it).

An *exporter* will `yield` serialized data given the interface method `export`. We choose to use `yield` as the output of this interface because it allows us to loop over this serialized data at a rate that does not surpass the limit that some integration channel may have for receiving data on the other end.

Exporters may do whatever they must to gather data, including possibly calling other APIs (course API, certificate API, etc.), but make a promise through the `export` method to `yield` some serialized data, to be used by a **transmitter**.

## Transmitters

A **transmitter** transmits exported (read: serialized) data to the integration channel using the API client. It uses the `transmit` interface method which takes a single `payload`, which is an `Exporter` object (so that it can be exported (serialized)). It then delegates the actual transmission of the payload to the API client.

## API Client

The **API Client** for an integration channel is the specialized client that knows how the integration channel expects to be tickled, typically through a RESTful interface.

The *API client* is responsible for, usually, making the RESTful call given a serialized payload that it receives from some *transmitter*. It will handle any authentication details that need to be dealt with for the given integration channel's interface.

## Models

Each integration channel may or may not require certain models. If it does not need to build its own, it may delegate any model requirements to the base integration channel models.

### Global Integration Channel Configuration

A **global integration channel configuration** defines some globally required configuration for a given integration channel, usually, for example, defining API paths, provider IDs, and so on. This configuration is then used throughout the integration channel's code, common to all Enterprise Customers that use this integration channel. 

### Integration Channel Enterprise Customer Configuration

An **integration channel enterprise customer configuration** is a enterprise customer-specific configuration model for the given integration channel; it defines some details that may differ among enterprise customers for the given integration channel.

### Catalog Data Transmission Audit

A **catalog data transmission audit** is a model used to keep track of course metadata that is transmitted to a given integration channel. It may differ according to the integration channel in question, so different integration channels may define their own.

### Learner Data Transmission Audit

A **learner data transmission audit** is a model used to keep track of learner data that is transmitted to a given integration channel. It may differ according to the integration channel in question, so different integration channels may define their own.

**JIRA:** [ENT-676](https://openedx.atlassian.net/browse/ENT-676)

**Merge deadline:** Before end of Sprint W.

**Testing instructions:**

1. Checkout the branch from https://github.com/edx/edx-platform/pull/16457 (`uman/degreed`), and make sure you're installing the `edx-enterprise`'s `uman/degreed` branch as well. Make sure migrations for the Degreed app run successfully, as well as for some of the others.
1. Set up an Enterprise Customer such that it's linked to some Catalog in Discovery, which contains some course(s) that are active, enrollable, and marketable (i.e. has a slug, is published, contains valid seats, has proper enrollment dates, etc..).
1. Setup a Degreed Global Configuration given the Degreed test environment creds (if you don't have them, ask in private).
1. Setup a Degreed Enterprise Customer Configuration, using any of the needed creds.
1. Take note of what service user is on your devstack for communicating with the discovery service (on Docker, it's `discovery_worker` by default).
1. SSH into the business sandbox, and as a normal user (not edxapp) run `sudo tail -f /edx/var/log/lms/edx.log`.

##### Course Metadata

1. Open a new terminal tab, SSH into the business sandbox, and run `./manage.py lms --settings=<environment> transmit_course_metadata --catalog_user <discovery_service_user> --channel DEGREED`.
1. Note that in the tab showing the logs, you get output that corresponds correctly to whatever set of courses you had available (oh, and unless you find a way to filter celery output, you'll see a bunch of other random logs if the sandbox is being used). If you get stuck and it seems like not all courses are being captured from discovery as they should, this is definitely a configuration issue (i.e. with whether courses are properly available, enrollable, etc.).

##### Learner Completion Data

1. SSO into Degreed using the Degreed account you have access to. You should do this by navigating to the course enrollment landing page for the Degreed enterprise you setup. You can use the edX demonstration course.
1. Enroll in the course successfully through the enterprise flow and then complete the course (just pass it at least, i.e. 60%+ grade).
1. Open a new terminal tab, SSH into the business sandbox, and run `./manage.py lms --settings=<environment> transmit_learner_data --api_user staff --channel DEGREED`.
1. Note that in the tab showing the logs, you get output that shows the learner data associated with the SSO details (i.e. email should be the Degreed email; you can see what it is by logging into Degreed, clicking your portrait, and clicking Settings, then going to the 'Email' tab).
1. Look on Degreed's side to see the course marked as completed.

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=232&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented:
This is a big WIP right now. Currently in the middle of fixing up the side-effects of refactoring a lot of the existing SAPSF code. The actual Degreed Integration code will be quite straightforward to add after that, because it's just about writing the API client (for the most part).

douglashall commented:
@UmanShahzad Nice write up! I think this would be great as a README somewhere in the repository.

douglashall commented:
@UmanShahzad Let me know how I can be most helpful here.

UmanShahzad commented:
@douglashall Yep, we can probably add it as a `README.md` in the `integrated_channels` folder, which contains all the integrated channel apps. I will definitely be expanding on it later to be more specific as I continue this PR.

In terms of help, right now I'm untangling code that is hard for even myself to keep track of alone. I'm relying on the existing tests to make sure I don't break anything (they've been really helpful thus far). But I'm not touching any existing *logic*, rather just where things go and what naming conventions they use (there seemed to be little standardization there, but that's going to be necessary to make it easy to add new channels, or else maintenance will be a headache), so the tests should keep things safe.

So when I get to implementing the Degreed work, the only help that'll be needed is help in thoroughly testing it. I'll ping you when I get there, thanks! :)

douglashall commented:
@UmanShahzad Sounds good. Thanks.

UmanShahzad commented:
Update: I have tests / quality passing for the SAPSF stuff (I moved around / deleted some tests though, so coverage isn't full, but I'll add that back at the end, after finishing Degreed Integration implementation).

Next up: Look at their API docs for a 2nd time, this time carefully, to determine how we'll do the data transform. Also will have to look into their API authentication (they don't want you requesting for an access token over and over -- you should wait for the expiration date on it, which is 60 days) and make sure the API client takes their requirements into account. Then there'll be smaller details like what to audit as part of the transmissions, etc..

Goal is to have a working version by the end of the sprint, reviewed (i.e. tested on the sandbox, at least). I will not focus on writing a test suite yet, nor getting coverage, until after we can see that the management commands work and we can see data get transferred to Degreed.

UmanShahzad commented:
@douglashall @itsjeyd Status update (FYI):

I've finished implementing Degreed's API Client that handles getting OAuth access tokens and actually making the requests. I've consequently updated all the Degreed-related configuration models to reflect what we'll actually need in the end (see `integrated_channels/degreed/models.py` if curious) for this integration.

Today I worked on completing Degreed's course metadata serializer (exporter), which 'transforms' edX course metadata from course discovery (really, from the Enterprise API's abstraction over it) into data that Degreed's API accepts and can understand. I *may* need some clarification on whether certain data points I chose are correct for what Degreed expects, but we can talk about that when polishing up, as the transformation for any data point is encapsulated into its own dedicated function, so making changes is easy (if curious, see `integrated_channels/degreed/exporters/course_metadata.py`).

Tomorrow I'll work on the learner completion status serializer and transmitter. Then some final touch-ups and local testing, and I can pass this over to you guys to review as a prototype (we can save code review for when we go head-on in adding a test suite and polishing up).

UmanShahzad commented:
I've gotten both commands implemented now.

UmanShahzad commented:
@itsjeyd @douglashall I've added test instructions for the course metadata transmission command.

Note that this PR not only renames the commands as part of another ticket, but makes the tasks asynchronous. I'm not 100% sure what to expect log-wise out of asynchronous tasks, because I tested these while they were still synchronous. I don't know where the logs go on devstacks.

UmanShahzad commented:
Ignore test failures for now -- they seem to not be running properly with the most up-to-date code.

UmanShahzad commented:
@douglashall @itsjeyd This should be able to go through an initial code review now

itsjeyd commented on integrated_channels/integrated_channel/transmitters/__init__.py:
```diff
@@ -0,0 +1,34 @@
+            * client - The REST API client that'll transmit serialized data.
+        """
+        self.enterprise_configuration = enterprise_configuration
+        self.client = client(enterprise_configuration) if client is not None else None
```
@UmanShahzad `if client is not None else None` seems a bit redundant? Is this just to make it explicit that `client` might be `None`? In that case we could consider adding an inline comment that mentions this, and just do `self.client = client(enterprise_configuration)` here.

UmanShahzad replied:
Yeah it's just to make sure we get an actual client. If I leave `client=None` here, it may seem a bit weird to then do just `client(enterprise_configuration)` to a reader. If I change it to always require a `client` as an argument, that may actually require a lot of other changes (since, from above, I wanted to keep a consistent interface if possible, and changing those will also require changing a bunch of tests, just for this).

I'll change it to `... if client else None` to make it a little less awkward sounding, though.

itsjeyd replied:
That sounds great, thanks for the details!

itsjeyd commented on integrated_channels/integrated_channel/transmitters/course_metadata.py:
```diff
@@ -16,85 +18,86 @@
-        return catalog_transmission_audit
+        Returns:
+            status_code (int): An integer status for the HTTP request
+            body (str): The SAP SuccessFactors server's response body
```
@UmanShahzad This should no longer reference SAPSF.

UmanShahzad replied:
:+1:

itsjeyd commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -0,0 +1,37 @@
+    It may be subclassed by specific integrated channel learner data transmitters for
+    each integrated channel's particular learner data transmission requirements and expectations.
+
+    TODO: Find any generic course metadata transmission logic across different integrated channels and put it here.
```
@UmanShahzad "Find any generic *learner data* transmission logic..."?

Do you plan on doing this in the scope of the current ticket?

UmanShahzad replied:
Ah yeah, bad copy and paste (again...)

I really would like not to (in the scope of this ticket). Would prefer to just get Degreed Integration going and do any optimizations or remaining refactors later.

itsjeyd replied:
Fine with me! My main intention was to draw attention to the `TODO`, and to see if you had any additional code coming :)

itsjeyd commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -0,0 +1,37 @@
+        """
+
+        :param payload:
+        :return:
```
@UmanShahzad This docstring needs a little bit of work :)

UmanShahzad replied:
Ouch, sorry about that :)

itsjeyd commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -0,0 +1,37 @@
+
+    def __init__(self, enterprise_configuration, client=IntegratedChannelApiClient):
+        """
+
```
@UmanShahzad Could you fill this in?

UmanShahzad replied:
:+1:

UmanShahzad commented on integrated_channels/README.md:
```diff
@@ -0,0 +1,4 @@
+# Integrated Channels
+
+Integrated channels are external systems that manage Enterprise Customer data. This package contains logic
+for configuring integrated channels and sending data between us and an integrated channel.
```
Not yet, actually. I'll do it at the end when wrapping up, because that's when we'll know 100% how everything is about to work. Otherwise I might have to come back here repeatedly to update that.

itsjeyd replied:
Sounds good!

UmanShahzad commented on integrated_channels/degreed/__init__.py:
```diff
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+"""
+Degreed module.
```
:+1:

UmanShahzad commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,153 @@
+        Send a completion status payload to the Degreed Completion Status endpoint
+
+        Args:
+            degreed_user_id (str): The degreed user id that the completion status is being sent for.
```
That's a mystery to me right now, too. We'll find out once SSO is setup :D But I assume it's just any old username -- in this case, it'd be a string. But the naming isn't the best. It's copied from `sapsf_user_id`.

UmanShahzad commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,153 @@
+        Raises:
+            HTTPError: if we received a failure response code from Degreed
+        """
+        url = self.global_degreed_config.degreed_base_url + self.global_degreed_config.completion_status_api_path
```
That's a good idea! :+1:

UmanShahzad commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,153 @@
+        Args:
+            url (str): The url to post to.
+            data (str): The json encoded payload to post.
+            scope (str): Must be one of
```
Haha, sorry about that. This is just one of the many docstrings that I forgot to get back to when I was in the jungle of the refactor.

itsjeyd replied:
Totally understandable! :D

UmanShahzad commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,153 @@
+        Raises:
+            HTTPError: If we received a failure response code from Degreed.
+            RequestException: If an unexpected response format was received that we could not parse.
+            ValueError: If the provided scope does not match any available Degreed API provider scopes.
```
Yep, that's old, sorry!

UmanShahzad commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,153 @@
+        response.raise_for_status()
+        data = response.json()
+        try:
+            return data['access_token'], datetime.datetime.utcfromtimestamp(data['expires_in'] + int(time.time()))
```
Sure, done!

UmanShahzad commented on integrated_channels/degreed/exporters/course_metadata.py:
```diff
@@ -0,0 +1,206 @@
+        """
+        Return a transformed version of course tags.
+
+        Note: none formal category tags for courses actually exist at this time. Update this when and if they do.
```
:+1:

UmanShahzad commented on integrated_channels/degreed/exporters/learner_data.py:
```diff
@@ -0,0 +1,62 @@
+            str(completed_date.year),
+            str(completed_date.month),
+            str(completed_date.day),
+        ]) if isinstance(completed_date, datetime) else None
```
Good idea!

UmanShahzad commented on integrated_channels/degreed/exporters/learner_data.py:
```diff
@@ -0,0 +1,62 @@
+            )
+        else:
+            LOGGER.debug(
+                'No learner data was sent for user "%s" because an Degreed user ID could not be found.',
```
Ah yeah, copy-and-paste from SAPSF. Thanks!

UmanShahzad commented on integrated_channels/degreed/models.py:
```diff
@@ -0,0 +1,273 @@
+from simple_history.models import HistoricalRecords
+
+from django.db import models
+from django.utils.encoding import python_2_unicode_compatible
```
Quality checks which include `isort` checks seem to be passing, so I'm not sure. Running `isort` again doesn't seem to change anything :\

itsjeyd replied:
Weird. Did you check if `isort` restores the current ordering if you change it manually to match what we'd expect?

UmanShahzad replied:
Just tested, and `isort` is forcing it back into the state it's in.

itsjeyd replied:
OK, thanks for checking!

UmanShahzad commented on integrated_channels/degreed/models.py:
```diff
@@ -0,0 +1,273 @@
+@python_2_unicode_compatible
+class DegreedEnterpriseCustomerConfiguration(EnterpriseCustomerPluginConfiguration):
+    """
+    The Enterprise specific configuration we need for integrating with Degreed.
```
:+1:

UmanShahzad commented on integrated_channels/degreed/models.py:
```diff
@@ -0,0 +1,273 @@
+    key = models.CharField(
+        max_length=255,
+        blank=True,
+        verbose_name="API Client ID",
```
Yeah, there is a mismatch, but it's mostly carry-over from the SAPSF integrated channel terminology used for the same thing. It took me some careful looking to see what `key` really meant, too. I can change it, but it might make the code harder to follow now between the 2 channels, and one of my goals has been to keep them quite consistent looking.

If you have some other idea or efficient way of doing this, do let me know.

itsjeyd replied:
> I can change it, but it might make the code harder to follow now between the 2 channels, and one of my goals has been to keep them quite consistent looking.

I'm 👍 on that. No need to make this particular PR bigger than it already is!

UmanShahzad commented on integrated_channels/degreed/models.py:
```diff
@@ -0,0 +1,273 @@
+            'employeeId': self.degreed_user_id,
+            'id': self.course_id,
+            'completionDate': self.completed_timestamp,
+        }
```
This is another carry-over from the SAPSF version.

The issue is that, at least for course metadata exports, it's happening through the exporter itself, with all those `transform_x` functions -- I *really* like that way of doing things, as it makes way more sense from a higher level, and is easier to follow in the code.

However, for some strange reason, the learner data exporter was (by "was" I mean this is how it was all wired up before and still is, despite the refactorizations) 'exporting' data through use of the learner data audit model. I tried hard to find a way to untangle this and make things consistent with the other cleaner technique of serialization, but I quickly realized that would mean a *lot* of extra refactorizations, this time not to just move things around, but also to change a bunch of logic which may have bad side-effects and unforeseen bugs in the existing SuccessFactors code. Complete coverage wouldn't guarantee us success here, no pun intended.

So I think we can still do this, and I would prefer it, but it would delay the delivery by quite a bit, and increase testing complexity. The tradeoff for cleaning up this part of the architecture is just more time...

itsjeyd replied:
OK, thanks for explaining the reasoning behind the current implementation. I'd be fine with doing additional refactoring as part of another ticket -- to make sure that this happens, though, maybe add `TODO`s to the docstrings of these methods, and create a follow-up ticket *now*?

CC @douglashall -- final decision up to you, of course :)

UmanShahzad replied:
@itsjeyd @douglashall A separate ticket would be the most reasonable given how big this PR already is, yeah. It's not going to affect functionality, but it's more of a continuation of the refactor to match the ideal, which we'd ultimately prefer at some point.

I will add some TODOs, and wait for @douglashall 's response to make the ticket.

douglashall replied:
I agree, let's create a ticket describing the additional refactor and move on with what we have.

UmanShahzad replied:
@douglashall Great, I'll make a ticket soon!

UmanShahzad replied:
@douglashall @itsjeyd https://openedx.atlassian.net/browse/ENT-761 is made.

UmanShahzad replied:
P.S. at least for this case, I was able to make the serialize method return all the data needed now.

UmanShahzad commented on integrated_channels/degreed/transmitters/course_metadata.py:
```diff
@@ -0,0 +1,24 @@
+
+    def __init__(self, enterprise_configuration, client=DegreedAPIClient):
+        """
+        Ensure by default that the client used for Degreed Course Metadata transmission is ``DegreedAPIClient``.
```
I actually think the docstring is misleading. It's not really ensuring anything, so I can just remove that word.

I don't have a good reason to keep it configurable, but if I recall my original thoughts on this correctly, it was to maintain the same signature for all subclasses of the base transmitter. That can always be broken without issues of course, but I'd actually prefer to keep it if we can.

UmanShahzad commented on integrated_channels/degreed/transmitters/learner_data.py:
```diff
@@ -0,0 +1,79 @@
+
+    def __init__(self, enterprise_configuration, client=DegreedAPIClient):
+        """
+        Ensure that, by default, the client used for Degreed Learner Data transmission is ``DegreedAPIClient``.
```
^^

UmanShahzad commented on integrated_channels/degreed/transmitters/learner_data.py:
```diff
@@ -0,0 +1,79 @@
+            enterprise_enrollment_id = learner_data.enterprise_course_enrollment_id
+            if learner_data.completed_timestamp is None:
+                # The user has not completed the course, so we shouldn't send a completion status call
+                LOGGER.debug('Skipping in progress enterprise enrollment {}'.format(enterprise_enrollment_id))
```
Thanks! I actually went ahead and `sed`'d this change across the codebase.

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+
+class CourseExporter(Exporter):
+    """
+    Base class for course metadata exporters to implement a "send" method on.
```
Woops, not sure what I meant here originally.

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+
+    def __init__(self, user, enterprise_configuration):
+        """
+        Save the appropriate details for use elsewhere in the object.
```
`removed_courses_resolved` and how it works was also quite a mystery to me.

But yeah, at least as of right now, they seem stable -- none of the specific integrated channels override how they get instantiated, so we should be to document them in the docstring itself.

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+        """
+        Return blanket data by default.
+
+        If a course exporter requires data transformation, it should override this property.
```
Yeah, I don't know of a use-case myself right now, so I'll do that.

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+        purposes.
+
+        Note that these integrated channels are presumed to be handling updating their course
+        availability on upstream catalogs through some other technique, i.e. DELETE requests
```
Degreed in this case (yeah, that also means we need to implement DELETE requests for Degreed if we plan on explicitly marking courses in upstream catalogs as inactive (same goes for removing learner completions); I think this would essentially remove them from the catalog entirely, which I'm not sure about doing. cc @douglashall) If we do need to remove them, that should be pretty quick to add to what we already have.

I've reworded things to make this clearer.

itsjeyd replied:
> I've reworded things slightly to remove any confusion, but would like not to mention Degreed directly of course, to keep the concept generic.

Thanks! Yeah, I didn't mean to suggest that this docstring should mention Degreed/SAPSF; just wanted to confirm if my understanding of its contents was correct :)

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+        made to the integrated channel's API to remove a course from the upstream catalog.
+
+        An example of an integrated channel that may utilize this: SAPSF -- this function
+        would be overriden to generate a course metadata payload that embeds availability
```
:+1:

UmanShahzad commented on integrated_channels/integrated_channel/exporters/learner_data.py:
```diff
@@ -31,6 +35,24 @@ class BaseLearnerExporter(object):
+        """
+        # The Grades API and Certificates API clients require an OAuth2 access token,
+        #  so cache the client to allow the token to be reused. Cache other clients for
+        #  general reuse.
```
:+1:

UmanShahzad commented on integrated_channels/integrated_channel/management/commands/__init__.py:
```diff
@@ -59,6 +47,36 @@ def add_arguments(self, parser):
+            if filter_kwargs:
+                integrated_channels = integrated_channels.filter(**filter_kwargs)
+
+            # Gen the learner data to each integrated channel
```
I didn't change the logic here either actually, just moved it up above the proclaimed private functions. I have no clue what this means and why that's here though, so I'll just scratch it, because the code looks readable enough itself.

UmanShahzad commented on integrated_channels/integrated_channel/models.py:
```diff
@@ -38,49 +55,139 @@ def channel_code():
+            courseCompleted="true" if self.course_completed else "false",
+            completedTimestamp=self.completed_timestamp,
+            grade=self.grade,
+        )
```
Yep, this was the original SAPSF implementation of the issue I described above.

UmanShahzad commented on integrated_channels/integrated_channel/tasks.py:
```diff
@@ -0,0 +1,52 @@
+@task(track_started=True)
+def transmit_learner_data(username, channel_code, channel_pk):
+    """
+    Allows each enterprise customer's integrated channel to collect and transmit data within its own celery task.
```
Argh, I think that was always there. Not sure why it's only being caught in review now.

UmanShahzad commented on integrated_channels/integrated_channel/tasks.py:
```diff
@@ -0,0 +1,52 @@
+    """
+    api_user = User.objects.get(username=username)
+    integrated_channel = INTEGRATED_CHANNEL_CHOICES[channel_code].objects.get(pk=channel_pk)
+    integrated_channel.transmit_learner_data(api_user)
```
Not sure if intended, as it seems to have always been like that. Adding it now.

UmanShahzad replied:
It seems to have been this way because the code was never reached. See comment that I added above.

itsjeyd replied:
@UmanShahzad 

> See comment that I added above.

Which comment, exactly? At this point there are so many comments that it's a bit hard to tell :)

UmanShahzad replied:
Ah I meant comment in code:

```diff
+    # Note: learner data transmission code paths don't raise any uncaught exception, so we don't need a broad
+    # try-except block here.
```

itsjeyd replied:
@UmanShahzad Ah, OK :) Thanks for clarifying.

UmanShahzad commented:
@itsjeyd Done addressing / commenting on this initial review -- it was reall hefty, thanks!

@douglashall Would be great to have a 3rd pair of eyes on this, because it's big.

itsjeyd commented on integrated_channels/integrated_channel/exporters/learner_data.py:
```diff
@@ -239,7 +257,7 @@ def _collect_grades_data(self, enterprise_enrollment, course_details):
             grade = self.grade_passing
 
-        # Otherwise, the course is still in progress
+        # Otherwise, the course is still in-progress
```
@UmanShahzad This looks like a false positive to me :)

UmanShahzad replied:
Ah, yeah :p

itsjeyd commented on integrated_channels/integrated_channel/transmitters/course_metadata.py:
```diff
@@ -0,0 +1,113 @@
+
+    def __init__(self, enterprise_configuration, client=IntegratedChannelApiClient):
+        """
+        By default, use the interface integrated channel API client which raises an error if used.
```
@UmanShahzad Thanks for adding this highly confusing docstring ;) :P 

I stared at it for a couple minutes and I think I get it now, but perhaps there is a way to make it a bit more accessible? For example:

* Do something with "interface", e.g., replace it with "default" or "abstract" or something?
* At first glance it's not clear why the integrated channel API client would raise an error "if used".

UmanShahzad replied:
Yeah, I'll use 'abstract' as that's probably more in line with how we think about it in Python. Will handle the last point with some rewording too.

itsjeyd commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -0,0 +1,37 @@
+
+    def __init__(self, enterprise_configuration, client=IntegratedChannelApiClient):
+        """
+        By default, use the interface integrated channel API client which raises an error if used.
```
@UmanShahzad Same comment here about being slightly confused by this docstring.

UmanShahzad replied:
:+1:

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -0,0 +1,104 @@
+        Parse the provided course into the format natively supported by the provider.
+        """
+        LOGGER.info('Processing course with ID %s', course_run['key'])
+        LOGGER.debug('Parsing course for %s: %s', self.enterprise_customer, json.dumps(course_run, indent=4))
```
Yeah we can do that -- I changed it for everything, except where we use "course metadata". I've been actually using that a lot and even the command is called this, so I've left it that way.

UmanShahzad commented:
@itsjeyd I've addressed the review comments I saw so far again.

itsjeyd commented on tests/test_models.py:
```diff
@@ -1197,36 +1160,37 @@ def test_string_conversion(self, method):
+    @mock.patch('integrated_channels.integrated_channel.models.CourseTransmitter')
+    def test_get_course_data_transmitter_raises(self, mock_course_transmitter):
+        mock_course_transmitter.return_value = 'mock_course_transmitter'
+        assert self.config.get_course_data_transmitter() == 'mock_course_transmitter'
```
@UmanShahzad Docstrings for the four tests above would be nice.

UmanShahzad replied:
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_utilities.py:
```diff
@@ -882,14 +883,16 @@ def test_get_course_track_selection_url_raises_exception(self):
-            exporter.get_serialized_data()
+        mock_enterprise_configuration = mock.MagicMock()
+        exporter = CourseExporter(mock_user, mock_enterprise_configuration)
+        assert next(exporter.export()) == b'[]'
```
@UmanShahzad Not sure, but would it be worth adding a test for a scenario where `get_enterprise_course_runs` returns one or more course runs? Returning zero course runs (the scenario that this test covers) seems like a bit of an edge case.

UmanShahzad replied:
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
It looks like since this is just testing the base ``CourseExporter``, that export function is just:

```python
    def export(self):
        yield json.dumps(self.courses, sort_keys=True).encode('utf-8'), 'POST'  # I added the 'POST' recently and updated the test
```

Since we don't need to test that `json.dumps` works properly, I think this is actually okay. But I have no clue why this is in `tests/test_utilities.py`, so I'm moving it to the right place.

UmanShahzad replied:
Ah and luckily in the new place I'm moving it to, we have fake course data available easily, so I'll now use that data in this test.

UmanShahzad commented on integrated_channels/sap_success_factors/models.py:
```diff
@@ -215,60 +200,7 @@ def _payload_data(self):
-            # contactHours=empty_value,
-            # cpeHours=empty_value,
-            # instructorName=empty_value,
-            # comments=empty_value,
```
@itsjeyd Not sure if a permanent decision has been made or not, but having commented out code like this is just clutter -- if we decide there's business value in this, we can just add it back, and we have the git changeset history if we forget what the values were.

itsjeyd replied:
Sounds good.

UmanShahzad commented on integrated_channels/sap_success_factors/models.py:
```diff
@@ -185,13 +156,16 @@ class Meta:
-            self.sapsf_user_id,
-            self.course_id
+        return (
+            '<SapSuccessFactorsLearnerDataTransmissionAudit {transmission_id} for enterprise enrollment {enrollment}, '
```
:+1:

UmanShahzad commented on integrated_channels/sap_success_factors/models.py:
```diff
@@ -106,74 +102,49 @@ def channel_code():
+    enterprise_course_enrollment_id = models.PositiveIntegerField(blank=False, null=False)
     course_id = models.CharField(max_length=255, blank=False, null=False)
     course_completed = models.BooleanField(default=True)
-    completed_timestamp = models.BigIntegerField()  # we send a UNIX timestamp to SuccessFactors
```
Actually, no good reason it seems. Adding this back and with some better formatting.

UmanShahzad commented on integrated_channels/sap_success_factors/migrations/0010_move_audit_tables_to_base_integrated_channel.py:
```diff
@@ -0,0 +1,29 @@
+    ]
+
+    database_operations = [
+        # This table literally just moves apps, so just rename it to prefix it with `integrated_channel_`.
```
:+1:

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/learner_data.py:
```diff
@@ -0,0 +1,57 @@
+        course_completed = False
+        if completed_date is not None:
+            completed_timestamp = parse_datetime_to_epoch_millis(completed_date)
+            course_completed = is_passing
```
True, it's probably what caused some of our woes with testing SAPSF course completions before -- this code had a different interpretation of course completion than what we had in our minds at the time :)

itsjeyd replied:
It's great how refactorings like this can unearth these types of issues :)

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/learner_data.py:
```diff
@@ -0,0 +1,57 @@
+
+class SapSuccessFactorsLearnerExporter(LearnerExporter):
+    """
+    Base class for exporting learner completion data to integrated channels.
```
Yep, it seems the same with the Degreed one.

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/course_metadata.py:
```diff
@@ -0,0 +1,263 @@
+        return [{
+            'locale': self.transform_language_code(course_run.get('content_language')),
+            'value': course_run.get('title') or ''
+        }]
```
SAPSF is weird on a million levels, and I'm seriously glad Degreed had a cleaner API. Even moreso because I didn't have to do anything at all for SSO :) Just a straight up SAML IdP, no strings attached.

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/course_metadata.py:
```diff
@@ -0,0 +1,263 @@
+
+    def transform_content(self, course_run):
+        """
+        Return the transformed version of the course content.
```
Yeah, good point -- I've added some explanation in the docstring.

itsjeyd replied:
Perfect, thanks!

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/course_metadata.py:
```diff
@@ -0,0 +1,263 @@
+        """
+        Return the transformed version of the course price.
+        """
+        return []
```
I don't actually know, but it used this value before, so I continue to pass this back to prevent a regression in however communication with SAPSF takes place.

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/course_metadata.py:
```diff
@@ -0,0 +1,263 @@
+        """
+        Return the transformed version of the course revision number.
+        """
+        return 1
```
I don't actually know, but it used this value before, so I continue to pass this back to prevent a regression in however communication with SAPSF takes place.

UmanShahzad commented on integrated_channels/sap_success_factors/exporters/course_metadata.py:
```diff
@@ -0,0 +1,263 @@
+                }
+
+        for course_key, summary in previous_audit_summary.items():
+            # Add a course payload to self.courses so that courses no longer in the catalog are marked inactive.
```
:+1:

UmanShahzad commented on test_utils/factories.py:
```diff
@@ -242,29 +255,154 @@ class Meta(object):
+        """
+        Meta for ``EnterpriseCustomerPluginConfigurationFactory``.
+        """
+
```
Yeah it actually would, unfortunately. I can disable it but, meh, it might be good to keep for some longer meta classes.

itsjeyd replied:
Alright!

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,142 @@
+        SAPSuccessFactorsGlobalConfiguration.objects.create(
+            completion_status_api_path="",
+            course_api_path="",
+            oauth_api_path=""
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,142 @@
+        course_exporter_mock.resolve_removed_courses.return_value = {}
+        transmitter = course_metadata.SapSuccessFactorsCourseTransmitter(self.enterprise_config)
+        transmitter.transmit(course_exporter_mock)
+        self.send_course_import_mock.assert_called_with(json.dumps(self.payload))
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,142 @@
+        assert catalog_transmission_audit.enterprise_customer_uuid == self.enterprise_config.enterprise_customer.uuid
+        assert catalog_transmission_audit.total_courses == len(self.payload)
+        assert catalog_transmission_audit.status == '200'
+        assert catalog_transmission_audit.error_message == ''
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,142 @@
+        assert catalog_transmission_audit.enterprise_customer_uuid == self.enterprise_config.enterprise_customer.uuid
+        assert catalog_transmission_audit.total_courses == len(self.payload)
+        assert catalog_transmission_audit.status == '500'
+        assert catalog_transmission_audit.error_message == 'error occurred'
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
You're right, there's a lot of duplication. Will use `ddt`.

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,142 @@
+        assert catalog_transmission_audit.enterprise_customer_uuid == self.enterprise_config.enterprise_customer.uuid
+        assert catalog_transmission_audit.total_courses == len(self.payload)
+        assert catalog_transmission_audit.status == '200'
+        assert catalog_transmission_audit.error_message == ''
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_learner_data.py:
```diff
@@ -0,0 +1,127 @@
+        self.global_config = factories.SAPSuccessFactorsGlobalConfigurationFactory(
+            completion_status_api_path="",
+            course_api_path="",
+            oauth_api_path=""
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1: Ended up doing this through the factory.

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_learner_data.py:
```diff
@@ -0,0 +1,127 @@
+        self.send_completion_status_mock = send_completion_status_mock.start()
+        self.addCleanup(send_completion_status_mock.stop)
+
+    def test_transmit_already_sent(self):
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_learner_data.py:
```diff
@@ -0,0 +1,127 @@
+        )
+        transmitter = learner_data.SapSuccessFactorsLearnerTransmitter(self.enterprise_config)
+        transmitter.transmit(self.exporter())
+        self.send_completion_status_mock.assert_not_called()
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
You're right, that seems to be a repeat.

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_learner_data.py:
```diff
@@ -0,0 +1,127 @@
+        transmitter.transmit(self.exporter())
+        self.send_completion_status_mock.assert_not_called()
+
+    def test_transmit_success(self):
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_learner_data.py:
```diff
@@ -0,0 +1,127 @@
+        assert payload.status == '200'
+        assert payload.error_message == ''
+
+    def test_transmit_failure(self):
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_utils.py:
```diff
@@ -0,0 +1,41 @@
+        ('2011-01-01T00:00:00Z', '2011-01-01T00:00:00Z', False),
+        ('2015-01-01T00:00:00Z', '2017-01-01T00:00:00Z', True),
+        ('2018-01-01T00:00:00Z', '2020-01-01T00:00:00Z', False),
+        ('2018-01-01T00:00:00', '2020-01-01T00:00:00', False),
```
Will address this when we begin working on the tests ticket (there's definitely work that has to be done for both SAPSF/Degreed tests, and not just the latter).

UmanShahzad replied:
Not fully sure but I think it's to test things without UTC being indicated in the string.

UmanShahzad commented:
@itsjeyd Thanks a lot for the review! I've addressed everything above the `tests/` folder diff; I plan on more easily addressing those (by virtue of focus) when we begin the test-focused ticket, which will be next week. The reason is that I actually need to add coverage for SAPSF that was lost when refactoring, and it would just be easier to do all test-related stuff at once.

UmanShahzad commented:
@douglashall This is definitely ready now for a review from you as well, as I think @itsjeyd has given it a full run.

UmanShahzad commented:
@itsjeyd @douglashall I've also made the tasks successfully asynchronous! You can `sudo tail -f /edx/var/log/lms/edx.log` to see the celery task output while you run it in another window. I'll add that to the testing instructions.

I think I have to go in and make slight changes to the course runs that are currently available on the sandbox too, so that data is actually processed. I also spent like 1 hour figuring out why the course information wasn't updating properly given my changes on discovery, and it was because of caching. If you run into something weird, know it's probably this.

douglashall commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,213 @@
+        self.session = None
+        self.expires_at = None
+
+    def send_completion_status(self, payload):
```
Can we change the name here to `create_course_completion` to better align with the Degreed API? I know we probably won't ever need to create book completions for instance, but then again who knows.

UmanShahzad replied:
:+1: Good idea

UmanShahzad replied:
This would actually require me to change some SAPSF stuff too, btw, because a base client is calling certain methods.

douglashall commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,213 @@
+            self.COMPLETION_PROVIDER_SCOPE
+        )
+
+    def delete_completion_status(self, payload):
```
Same here about the name `delete_course_completion`.

UmanShahzad replied:
:+1:

douglashall commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,213 @@
+            self.COMPLETION_PROVIDER_SCOPE
+        )
+
+    def send_course_import(self, payload):
```
More naming alignment `create_course_content`.

UmanShahzad replied:
:+1:

douglashall commented on integrated_channels/degreed/client.py:
```diff
@@ -0,0 +1,213 @@
+            self.CONTENT_PROVIDER_SCOPE
+        )
+
+    def delete_course_import(self, payload):
```
`delete_course_content`

UmanShahzad replied:
:+1:

UmanShahzad commented:
@douglashall Thanks, I changed 'em -- will begin tests next week :)

UmanShahzad commented:
Codecov is pretending we met coverage requirements. Yeah right.

UmanShahzad commented:
Rebased on `master` from e6c4e5921695e947162d15ca449cf0274077d5fa.

douglashall commented on integrated_channels/integrated_channel/tasks.py:
```diff
@@ -52,18 +52,16 @@ def transmit_learner_data(username, channel_code, channel_pk):
-            )
+        LOGGER.exception(
+            'Transmission of learner data failed for user "%s" and for integrated '
+            'channel with code "%s" and id "%s".', username, channel_code, channel_pk
```
Can we standardize on brackets `[]` instead of double-quotes `""` for wrapping log values?

UmanShahzad replied:
:+1:

UmanShahzad commented:
This PR just got way bigger... the test code was *a lot*, and I'm not 100% yet, nor have I addressed the test-related comments.

UmanShahzad commented:
Finally got 100% coverage, phew...

@douglashall @itsjeyd I will update the documentation later. If it's not urgent, we can make a 2 point story for me to finish it up next sprint.

@itsjeyd I haven't yet addressed your comments above for the tests, so I will do that my tomorrow morning.

itsjeyd commented:
@UmanShahzad Thanks for the updates. Personally I'd be fine with a separate story for finishing up the documentation.

CC @douglashall

UmanShahzad commented:
@douglashall @itsjeyd Ready for review! Will put on sandbox.

UmanShahzad commented:
Latest changes on the sandbox.

itsjeyd commented on tests/test_models.py:
```diff
@@ -1168,21 +1168,33 @@ def test_channel_code_raises(self):
     @mock.patch('integrated_channels.integrated_channel.models.CourseTransmitter')
     def test_get_course_data_transmitter_raises(self, mock_course_transmitter):
+        """
+        The configuration returns the appropriate course transmitter,
```
@UmanShahzad Typo: "," should be ".", for this test and three tests above.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_course_metadata.py:
```diff
@@ -45,89 +48,90 @@ def setUp(self):
+                    'in_catalog': True,
+                    'status': 'ACTIVE',
+                }
+            }
```
@UmanShahzad Do we need to define this again, or could we just reference `audit_summary` that gets passed in from `ddt`?

UmanShahzad replied:
Ah sorry, nice catch, this shouldn't be here :+1:

itsjeyd commented on tests/test_integrated_channels/test_degreed/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,113 @@
+        # Add a course marked for deletion by giving it just the single key; the course run ID.
+        exporter.courses.append({'contentId': 'course-id-for-deletion'})
+        data = exporter.export()
+        json_dump1, method1 = (
```
@UmanShahzad It would be nice to be a bit more specific here (and in L62 below). How about something like `json_dump_deleted, delete_method` (and `json_dump_posted, post_method`)?

In any case, you could update the inline comments ("JSON Dump 1", "Method 1", etc.) to be a bit more specific.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_degreed/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,113 @@
+    def test_transform_description_returns_empty_str(self):
+        course_run = {'full_description': 'a' * (DegreedCourseExporter.LONG_STRING_LIMIT + 1)}
+        exporter = DegreedCourseExporter('fake-user', self.config)
+        assert exporter.transform_description(course_run) == ''
```
@UmanShahzad Let's add docstrings to the three tests above.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_degreed/test_exporters/test_learner_data.py:
```diff
@@ -0,0 +1,96 @@
+from integrated_channels.degreed.exporters.learner_data import DegreedLearnerExporter
+from pytest import mark
+
+from django.utils import timezone
```
@UmanShahzad I might be duplicating an earlier comment here, but would `isort` put this above `from freezegun import freeze_time`?

UmanShahzad replied:
Yeah I always run `isort` on the code, and this is what it produces. Changing it makes `isort` complain and calling it again after the change puts it back to how it is.

:+1:

itsjeyd replied:
@UmanShahzad Weird, but OK ;)

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,47 @@
+        self.mock_ent_courses_api_with_pagination(
+            enterprise_uuid=str(self.enterprise_customer.uuid),
+            course_run_ids=['course-v1:edX+DemoX+Demo_Course_1']
+        )
```
@UmanShahzad Same note here about moving this down to keep mock definitions together.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,64 @@
+            sapsf_user_id="user_id",
+            secret="client_secret",
+        )
+        self.global_config = factories.SAPSuccessFactorsGlobalConfigurationFactory()
```
@UmanShahzad This code isn't part of the `tests/test_integrated_channels/test_sap_success_factors/` package; should it be referencing SAP-specific classes?

UmanShahzad replied:
We need some non-abstract configuration for these things to work, so it's okay for it be to any arbitrary channel (atm between SAPSF/Degreed).

itsjeyd replied:
@UmanShahzad OK, could this be mentioned in a docstring (at class or method level)?

UmanShahzad replied:
Yep! :+1:

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,64 @@
+        )
+        self.global_config = factories.SAPSuccessFactorsGlobalConfigurationFactory()
+
+        # Mocks
```
@UmanShahzad Putting mocks in a separate "section" -- nice! :)

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,64 @@
+        )
+        self.delete_course_content_mock = delete_course_content_mock.start()
+        self.delete_course_content_mock.return_value = 200, '{}'
+        self.addCleanup(delete_course_content_mock .stop)
```
@UmanShahzad Remove extra space before `.stop`.

UmanShahzad replied:
Woops, copy paste fail. :+1:

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py:
```diff
@@ -0,0 +1,64 @@
+        """
+        transmitter = CourseTransmitter(self.enterprise_config)
+        transmitter.transmit_block('fake-course-block', method='DELETE')
+        self.delete_course_content_mock.assert_called_once_with('fake-course-block')
```
@UmanShahzad Should this module also include a [`test_transmit_block_post`](https://github.com/edx/edx-enterprise/blob/uman/degreed/integrated_channels/integrated_channel/transmitters/course_metadata.py#L95-L96) method? If not, I think you could remove the definition of [`create_course_content_mock`](https://github.com/edx/edx-enterprise/blob/uman/degreed/tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py#L37-L42) from `setUp` above.

UmanShahzad replied:
:+1: I've used `ddt` to also use `POST`, and the `create_course_content_mock`.

itsjeyd replied:
@UmanShahzad Nice!

itsjeyd commented on tests/test_integrated_channels/test_sap_success_factors/test_client.py:
```diff
@@ -242,7 +233,7 @@ def test_expired_access_token(self):
-        assert len(responses.calls) == 3  # pylint: disable=no-member
-        assert responses.calls[0].request.url == self.url_base + self.oauth_api_path  # pylint: disable=no-member
-        assert responses.calls[1].request.url == self.url_base + self.oauth_api_path  # pylint: disable=no-member
-        assert responses.calls[2].request.url == self.url_base + self.course_api_path  # pylint: disable=no-member
```
@UmanShahzad Thanks for cleaning up these `pylint` exceptions!

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_sap_success_factors/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,62 @@
+        self.mock_ent_courses_api_with_pagination(
+            enterprise_uuid=str(self.enterprise_customer.uuid),
+            course_run_ids=['course-v1:edX+DemoX+Demo_Course_1']
+        )
```
@UmanShahzad Same note here about moving this down to keep mock definitions together.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_integrated_channels/test_sap_success_factors/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,62 @@
+            exporter.transform_language_code('this-is-incomprehensible')
+        assert str(exc_info.value) == (
+            'Language codes may only have up to two components. Could not parse: this-is-incomprehensible'
+        )
```
@UmanShahzad These tests look good, let's just add docstrings.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -120,47 +149,80 @@ def test_transmit_course_metadata_task_with_error(
+            '"url": "'+settings.LMS_ROOT_URL+'/enterprise/'+enterprise_uuid_for_success+'/course/'
+            ''+course_run_id_for_success+'/enroll/", "videoUrl": ""}], '
+            '"orgCode": "Degreed Company", "providerCode": "DEGREED"}'
+        )
```
@UmanShahzad Not sure if `pylint` would complain, but putting spaces around `+` would improve readability here, I think.

UmanShahzad replied:
It won't, so yes, sure :) :+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -82,32 +115,28 @@ def test_override_user(self):
-            get_oauth_access_token_mock,
-    ):
+            sapsf_create_course_content_mock,
+            sapsf_get_oauth_access_token_mock,
```
@UmanShahzad How come it's not necessary to mock `DegreedAPIClient._get_oauth_access_token` here?

UmanShahzad replied:
Because the OAuth calls are made through the interface methods, i.e. `create_course_content`, so mocking them is sufficient. Not necessarily the case with SAPSF :p

itsjeyd replied:
@UmanShahzad Ahhh... right :P

itsjeyd commented on tests/test_management.py:
```diff
@@ -171,18 +233,21 @@ def test_transmit_course_metadata_task_with_error(
-            get_oauth_access_token_mock,
-    ):
+            sapsf_create_course_content_mock,
+            sapsf_get_oauth_access_token_mock,
```
@UmanShahzad Same question here about mocking `DegreedAPIClient._get_oauth_access_token`.

UmanShahzad replied:
(see above) :+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -197,37 +262,64 @@ def test_transmit_course_metadata_task_success(
+            '"title": "edX Demonstration Course", "url": "'+settings.LMS_ROOT_URL+'/enterprise/'
+            ''+uuid+'/course/'+course_run_ids[1]+'/enroll/", '
+            '"videoUrl": ""}], "orgCode": "Degreed Company", "providerCode": "DEGREED"}'
+        )
```
@UmanShahzad Same question here about putting spaces around `+`.

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -257,8 +351,7 @@ def test_transmit_course_metadata_task_no_channel(self):
         """
-        Test the data transmission task with enterprise customer which have no
-        course catalog.
+        Test the data transmission task with enterprise customer which have no course catalog.
```
@UmanShahzad Typo: "which have" should be "that has" (or "who has")

UmanShahzad replied:
:+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -514,23 +653,42 @@ class TestLearnerDataTransmitIntegration(unittest.TestCase):
+        )
+        self.degreed_create_course_completion = degreed_create_course_completion.start()
+        self.degreed_create_course_completion.return_value = 200, '{}'
+        self.addCleanup(degreed_create_course_completion.stop)
```
@UmanShahzad Similar question here about mocking `DegreedAPIClient._get_oauth_access_token`.

UmanShahzad replied:
(see above) :+1:

itsjeyd commented on tests/test_management.py:
```diff
@@ -514,23 +653,42 @@ class TestLearnerDataTransmitIntegration(unittest.TestCase):
+    def setUp(self):
+        super(TestLearnerDataTransmitIntegration, self).setUp()
+
+        # pylint: disable=invalid-name
```
@UmanShahzad Could you just disable `invalid-name` for the method as a whole? Then you wouldn't need to re-enable it in L681 below.

UmanShahzad replied:
:+1:

UmanShahzad replied:
Ok turns out I couldn't do it -- or at least I don't know how to scope it to just a method. If you've any idea I'll make the change -- google isn't helping here :p

itsjeyd replied:
@UmanShahzad Hm, I guess I was assuming that putting the directive (is that what these lines are called?) right above the method definition would do it... If that's not the case then I'm not sure. Also tried googling this, to no avail. (Which makes me think we're missing something -- what we're trying to do doesn't seem super esoteric to me, so perhaps the reason why nobody else seems to be having problems with it is that it's actually super easy to accomplish? :))

itsjeyd replied:
@UmanShahzad In any case, I'm fine with keeping this. No need to spend a lot of time trying to figure out how to change it.

UmanShahzad replied:
@itsjeyd There've been a lot of times I would like to control the scope of pylint more carefully (i.e., disable pylint for the next *5* lines only, etc.), but I think these features simply don't exist. It sounds like a good opportunity to contribute something ;)

itsjeyd commented on tests/test_models.py:
```diff
@@ -1309,13 +1314,75 @@ def test_string_conversion(self, method):
-        expected_to_str = "<SAPSuccessFactorsGlobalConfiguration with id 2>"
+        expected_to_str = (
+            "<DegreedLearnerDataTransmissionAudit 1 for enterprise enrollment 5, Degreed user degreed_user,"
+            " and course course-v1:edX+DemoX+DemoCourse>"
```
@UmanShahzad Nit: Include space at the end of the previous line.

UmanShahzad replied:
:+1:

UmanShahzad commented on integrated_channels/degreed/models.py:
```diff
@@ -254,23 +254,28 @@ def __repr__(self):
         # is serialized, remove the serialization here and make the learner data exporter handle the work.
         """
-        return json.dumps(self._payload_data(), sort_keys=True)
+        return json.dumps(self._payload_data(getattr(args[0], 'degreed_company_id')), sort_keys=True)
```
Sounds good :+1: After seeing this again I realized it's cleaner to make this a keyword argument instead anyway.

UmanShahzad commented on integrated_channels/integrated_channel/exporters/course_metadata.py:
```diff
@@ -40,15 +40,15 @@ def __init__(self, user, enterprise_configuration):
+        Export the set of courses generically by encoding them in JSON format, plus the HTTP method to use.
         """
-        yield json.dumps(self.courses, sort_keys=True).encode('utf-8')
+        yield json.dumps(self.courses, sort_keys=True).encode('utf-8'), 'POST'
```
In order to allow Degreed's course metadata transmitter to just re-use the base integrated channel transmitter's logic (I think SAPSF was already doing so), I had to make the base one use a configurable HTTP method, since Degreed Integration should be able to POST or DELETE depending on the course's enrollability. It should be a good interface to follow for future integrations that will also use methods besides just POST, like SAPSF was.

UmanShahzad replied:
Oh, and the reason it came up now (good thing it did) and not before, is because I noticed I'd have to rewrite almost the same exact test code for Degreed that exists for SAPSF to get full coverage in its course metadata transmitter, which didn't feel right of course.

itsjeyd replied:
@UmanShahzad Sounds good, thanks for the detailed answers!

UmanShahzad commented on integrated_channels/integrated_channel/migrations/0004_catalogtransmissionaudit_channel.py:
```diff
@@ -0,0 +1,19 @@
+        migrations.AddField(
+            model_name='catalogtransmissionaudit',
+            name='channel',
+            field=models.CharField(default='SAP', max_length=30),
```
Good point and yeah I'm not sure what it should default to.

And I think you're right that we should make it a choice out of existing channels, but at the same time this wouldn't be visible in the admin, so it wouldn't be that useful. What do you think we should do? cc @douglashall

douglashall replied:
Will migrations not work if we don't define a default?

UmanShahzad replied:
I realized there's an option `preserve_default` for `AddField` that we can set to `False` and have this as a one-off for all existing rows. The one-off default can be `'SAP'` because all existing rows are for that, and then this can carry on without one; since the field is non-nullable, new rows will require their own channel, i.e. `'DEGREED'`.

itsjeyd replied:
@UmanShahzad 

> ... there's an option `preserve_default` for `AddField` that we can set to `False` and have this as a one-off for all existing rows.

\#TIL

UmanShahzad commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -28,10 +33,52 @@ def __init__(self, enterprise_configuration, client=IntegratedChannelApiClient):
+            payload: The learner completion data payload to send to the integrated channel.
+            kwargs: Contains integrated channel-specific information for customized transmission variables.
+                - app_label: The app label of the integrated channel for whom to store learner data records for.
+                - model_name: The name of the specific learner data record model to use.
```
Yes it should! :+1:

UmanShahzad commented on integrated_channels/integrated_channel/transmitters/learner_data.py:
```diff
@@ -28,10 +33,52 @@ def __init__(self, enterprise_configuration, client=IntegratedChannelApiClient):
+        )
+        for learner_data in payload.export():
+            serialized_payload = learner_data.serialize(self.enterprise_configuration)
+            LOGGER.info(serialized_payload)
```
:+1:

UmanShahzad commented on integrated_channels/integrated_channel/client.py:
```diff
@@ -21,6 +21,22 @@ class IntegratedChannelApiClient(object):
+            enterprise_configuration: An enterprise customers's configuration model for connecting with the channel
+
+        Raises:
+            ValueError: If an enterprise configuration is not provided..
```
:+1:

UmanShahzad commented on test_utils/factories.py:
```diff
@@ -357,9 +340,11 @@ class Meta(object):
     completion_status_api_path = factory.LazyAttribute(lambda x: FAKER.file_path())
     course_api_path = factory.LazyAttribute(lambda x: FAKER.file_path())
     oauth_api_path = factory.LazyAttribute(lambda x: FAKER.file_path())
+    provider_id = 'SAP'
```
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_client.py:
```diff
@@ -0,0 +1,287 @@
+        assert len(responses.calls) == 2
+        assert responses.calls[0].request.url == self.oauth_url
+        assert responses.calls[1].request.url == self.course_url
+
```
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_client.py:
```diff
@@ -0,0 +1,287 @@
+    @responses.activate
+    def test_expired_token(self):
+        """
+        If our token expires after some call, make sure to get it again.
```
Well time is actually frozen for these tests, so if it expires in > 0 time it will never actually expire and we couldn't test token expiration. So in this test we're basically saying "Make a call, have the token expire after waiting some time (technically no time), and make a call again and notice 2 OAuth calls are required". And this contrasts with other tests where we don't need to make 2 OAuth calls, in particular the very next one, '`test_existing_token_is_valid`'.

itsjeyd replied:
@UmanShahzad Ah, I see. I think it would be great to have this info available when reading the code, so how about putting it in the method's docstring?

UmanShahzad replied:
Sounds good, :+1:

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_client.py:
```diff
@@ -0,0 +1,287 @@
+    @responses.activate
+    def test_oauth_response_missing_keys(self):
+        """
+        A ``requests.RequestException`` is raised for when the call for an OAuth2 access token fails.
```
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_client.py:
```diff
@@ -0,0 +1,287 @@
+            responses.POST,
+            self.oauth_url,
+            json={},
+            status=200
```
I should change the docstring here to just say "A ``requests.RequestException`` is raised when the call for an OAuth2 access token returns no data."

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_client.py:
```diff
@@ -0,0 +1,287 @@
+
+        degreed_api_client = DegreedAPIClient(self.enterprise_config)
+        with self.assertRaises(requests.RequestException):
+            degreed_api_client.delete_course_content({})
```
Nope, it shouldn't change what we're testing for :+1:

itsjeyd replied:
@UmanShahzad Agreed :D

Cf. [this conversation](https://github.com/edx/edx-enterprise/pull/232#discussion_r152239358)

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,113 @@
+        self.mock_ent_courses_api_with_pagination(
+            enterprise_uuid=str(self.enterprise_customer.uuid),
+            course_run_ids=['course-v1:edX+DemoX+Demo_Course_1']
+        )
```
:+1:

UmanShahzad commented on tests/test_integrated_channels/test_degreed/test_exporters/test_course_metadata.py:
```diff
@@ -0,0 +1,113 @@
+        """
+        exporter = DegreedCourseExporter('fake-user', self.config)
+        # Add a course marked for deletion by giving it just the single key; the course run ID.
+        exporter.courses.append({'contentId': 'course-id-for-deletion'})
```
Yeah that's a little confusing, but it's actually done through `self.mock_ent_courses_api_with_pagination` in the setup: it already adds a course with id `course-v1:edX+DemoX+Demo_Course_1` to the data returned for when the exporter is initialized. I don't like magic but ah well... I've added a comment to take note of it.

UmanShahzad commented:
@itsjeyd Alright, I addressed your review, thanks for doing it!

UmanShahzad commented:
I'm really confused about why that 1 test is failing. It's not failing locally...

UmanShahzad commented:
Builds are definitely bugging out, considering 1 build passed and another almost completely failed. Not sure what's up with it but I'm just restarting and hoping for the best.

UmanShahzad commented:
@itsjeyd Thanks so much! I've addressed things now, and rebased on `master` from ac19f8207746f941df0454fefb68df2a8e6e03d6. The rebase actually fixes the test issues AFAIK.

UmanShahzad commented:
@itsjeyd Ready for testing the learner data transmission command now -- I've also updated test instructions. Let me know if anything's vague.

itsjeyd commented on tests/test_integrated_channels/test_integrated_channel/test_transmitters/test_course_metadata.py:
```diff
@@ -25,6 +25,8 @@ class TestCourseTransmitter(unittest.TestCase):
         super(TestCourseTransmitter, self).setUp()
         enterprise_customer = factories.EnterpriseCustomerFactory(name='Starfleet Academy')
+        # We need some non-abstract configuration for these things to work,
+        # so it's okay for it be to any arbitrary channel. We randomly choose SAPSF.
```
@UmanShahzad Ack! There's a typo here: "be to" should be "to be".

Be to or not be to, that is the question! (Sorry, couldn't resist 😎)

UmanShahzad replied:
:sunglasses: :sunglasses: :sunglasses:

UmanShahzad commented:
@itsjeyd Thanks a lot!! @douglashall can you look through / test? This is ready to ship after your call.

itsjeyd commented:
@UmanShahzad No worries :) Will be happy to see this merge!

UmanShahzad commented:
Don't merge yet -- there's one tiny thing to fix up.

UmanShahzad commented:
Ready for merge :)
